### PR TITLE
[CDAP-14756] Refactor all APIs to use Metadata SPI directly

### PR DESCRIPTION
**Pull request by anew**
... and get rid of old MetadataStore classes. The HTTP endpoints remain backwards-compatible through an optional flag that is the default for now. In a follow-up, I will flip that default and change the metadata client and CLI classes to expect new result formats. 

Commits are (almost) self-contained and self-describing and best reviewed one by one.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java [hunk 1]
```diff
@@ -120,41 +104,10 @@ public Lineage computeLineage(final DatasetId sourceDataset, long startMillis, l
...
-  public Lineage computeLineage(final DatasetId sourceDataset, long startMillis, long endMillis,
-                                int levels) throws NotFoundException {
+  public Lineage computeLineage(final DatasetId sourceDataset, long startMillis, long endMillis, int levels) {
     return doComputeLineage(sourceDataset, startMillis, endMillis, levels, null);
   }
 
-  /**
-   * @return metadata associated with a run
-   */
-  public Set<MetadataRecord> getMetadataForRun(ProgramRunId run) {
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java:131]: Is this removed because it is not being used anywhere?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java [hunk 2]
```diff
@@ -112,7 +123,7 @@
   MetadataSubscriberService(CConfiguration cConf, MessagingService messagingService,
                             DatasetFramework datasetFramework, TransactionSystemClient txClient,
                             MetricsCollectionService metricsCollectionService,
-                            MetadataStore metadataStore,
+                            AuditMetadataStorage metadataStorage,
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java]: Rather than letting services individually decide whether they want audit support or not, I think it would be cleaner if  `AuditMetadataStorage` is returned by `MetadataStorageProvider`. This way if we want to turn off auditing, then we can do it in one place.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java [hunk 3]
```diff
@@ -78,18 +83,19 @@
...
 
-  private final MetadataStore metadataStore;
+  private final MetadataStorage metadataStorage;
   private final NamespaceTable namespaceTable;
   private final AppMetadataStore appMetadataStore;
   private final ProgramScheduleStoreDataset scheduleDataset;
   private final PreferencesTable preferencesTable;
 
-  public ProfileMetadataMessageProcessor(MetadataStore metadataStore, StructuredTableContext structuredTableContext) {
+  public ProfileMetadataMessageProcessor(MetadataStorage metadataStorage,
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java:92]: Don't the metadata updates that happen here need to be published for audit?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java [hunk 4]
```diff
@@ -92,19 +94,21 @@ public void testWorkflowTags() throws Exception {
...
     Assert.assertTrue("Workflows should not be tagged with fork node names, but found the following fork nodes " +
                         "in the workflow's system tags: " + intersection, intersection.isEmpty());
 
     // verify that metadata was added for the workflow's schedule. First wait for the metadata to show up
-    Tasks.waitFor(false, () -> metadataStore.getProperties(
-      MetadataScope.SYSTEM, appId.toMetadataEntity()).isEmpty(), 5, TimeUnit.SECONDS);
-    Map<String, String> metadataProperties = metadataStore.getMetadata(MetadataScope.SYSTEM,
-                                                                       appId.toMetadataEntity()).getProperties();
+    Tasks.waitFor(false, () -> metadataStorage.read(new Read(appId.toMetadataEntity(),
+                                                             MetadataScope.SYSTEM, MetadataKind.TAG)).isEmpty(),
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java]: Typo: the old code is reading properties not tags

**Review by poornachandra — COMMENTED** [review]
@anew I have reviewed the first 5 commits, and have a few comments. I'll continue the review later

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java (hunk 1 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java:131]: correct. It was decided in 5.x to deprecate it and in 6.0 we removed support for this. The UI never used it anyway.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java (hunk 2 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java]: good point

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java (hunk 3 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java:92]: +1

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java (hunk 4 above)
Comment (anew) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java]: oops. will fix.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java (hunk 3 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java:92]: actually they are. Because the MetadataStorage here is not injected, but passed in through the constructor in MetadataMessageProcessor, which uses the Auditing one.

**Comment by anew** [comment]
@poornachandra I force-pushed because I had to resolve conflicts after addressing your first set of comments. Did not change any code for that, so please continue after the 5th commit.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java [hunk 5]
```diff
@@ -1020,14 +967,6 @@ public void testInvalidSearchParams() throws Exception {
     } catch (BadRequestException e) {
       // expected
     }
-
-    // search with invalid query
-    try {
-      searchMetadata(NamespaceId.DEFAULT, "");
```
Comment (poornachandra) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1026]: Why is this removed?

**Review by poornachandra — APPROVED** [review]
@anew Just one comment

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java (hunk 5 above)
Comment (anew) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1026]: Because it is no longer invalid. The existing code has "" as the default in the REST API, but that never worked. So i changed it mean the same as "*".

**Comment by anew** [comment]
Build was successful: https://builds.cask.co/browse/CDAP-DUT6859
Verification Build also passed: https://builds.cask.co/browse/CDAP-DBV20/latest
Running one more mini ITN after the rebase.

**Comment by anew** [comment]
Mini ITN: https://builds.cask.co/browse/IT-ITM-121

**Comment by anew** [comment]
ITM passed, squashing and merging now.
